File: solutions/cpp/0025.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (!head || k == 1) return head;

        ListNode* dummy = new ListNode(NULL);
        dummy->next = head;

        int len = 0;
        for (auto curr = head; curr; curr = curr->next) len++;

        auto prev = dummy;
        auto curr = head;
        for (int i = 0; i < len / k; i++) {
            for (int j = 0; j < k - 1; j++) {
                auto next = curr->next;
                curr->next = next->next;
                next->next = prev->next;
                prev->next = next;
            }
            prev = curr;
            curr = curr->next;
        }

        return dummy->next;
    }
};

File: solutions/python3/0025.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head or k == 1:
            return head

        dummy = ListNode(None)
        dummy.next = head

        len = 0
        curr = head
        while curr:
            len += 1
            curr = curr.next

        prev = dummy
        curr = head
        for i in range(len // k):
            for j in range(k - 1):
                next = curr.next
                curr.next = next.next
                next.next = prev.next
                prev.next = next
            prev = curr
            curr = curr.next

        return dummy.next
